internal/planner/telemetry.go

```diff
@@ -0,0 +1,127 @@
+func must[T any](obj T, err error) T {
+	if err != nil {
+		panic(err)
+	}
+	return obj
+}
+
+var (
+	pkg                  = "github.com/canonical/github-runner-operators/internal/planner"
+	meter                = otel.Meter(pkg)
+	logger               = telemetry.NewLogger(pkg)
+	flavorPressureMetric = must(
+		meter.Int64ObservableGauge(
+			flavorPressureMetricName,
+			metric.WithDescription("The current pressure value for each flavor"),
+			metric.WithUnit("{pressure}"),
+		),
+	)
+)
+
+// Metrics encapsulates all OpenTelemetry metrics for the planner service.
+type Metrics struct {
+	mu    sync.RWMutex
+	store FlavorStore
+}
+
+// NewMetrics initializes the planner metrics and registers the observable gauge.
+func NewMetrics(store FlavorStore) *Metrics {
+	m := &Metrics{
+		store: store,
+	}
+
+	must(
+		meter.RegisterCallback(
+			m.collectFlavorPressure,
+			flavorPressureMetric,
+		),
+	)
+
+	return m
+}
+
+// collectFlavorPressure collects flavor pressure metrics for all platforms.
+func (m *Metrics) collectFlavorPressure(ctx context.Context, observer metric.Observer) error {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+
+	// Pressure value is read directly from database at collection time
+	// Performance should not be an issue since metrics scraping typically occurs every 5-60 seconds.
+	platforms := []string{flavorPlatform}
+	for _, platform := range platforms {
+		m.collectPlatformPressure(ctx, observer, platform)
+	}
+	return nil
+}
+
+// collectPlatformPressure collects flavor pressure metrics for a single platform.
+func (m *Metrics) collectPlatformPressure(ctx context.Context, observer metric.Observer, platform string) {
+	flavors, err := m.store.ListFlavors(ctx, platform)
+	if err != nil {
+		logger.ErrorContext(ctx, "failed to list flavors for metrics", "platform", platform, "error", err)
+		return
+	}
+
+	if len(flavors) == 0 {
+		return
+	}
+
+	flavorNames := extractFlavorNames(flavors)
+	pressures, err := m.store.GetPressures(ctx, platform, flavorNames...)
+	if err != nil {
+		logger.ErrorContext(ctx, "failed to get pressures for metrics", "platform", platform, "error", err)
+		return
+	}
+
+	m.observeFlavorPressures(observer, platform, flavors, pressures)
+}
+
+// extractFlavorNames extracts the names from a list of flavors.
+func extractFlavorNames(flavors []database.Flavor) []string {
+	names := make([]string, 0, len(flavors))
+	for _, f := range flavors {
+		names = append(names, f.Name)
+	}
+	return names
+}
+
+// observeFlavorPressures records pressure observations for each flavor.
+func (m *Metrics) observeFlavorPressures(observer metric.Observer, platform string, flavors []database.Flavor, pressures map[string]int) {
+	for _, f := range flavors {
+		v := pressures[f.Name]
+		observer.ObserveInt64(
+			flavorPressureMetric,
+			int64(v),
+			metric.WithAttributes(
+				attribute.String("platform", platform),
+				attribute.String("flavor", f.Name),
+			),
+		)
+	}
+}
```

Comment:
It seems explicit unit tests are missing for these methods (though these methods are tested via the unit tests in server_test.go).